hphp/runtime/vm/jit/vasm-arm.cpp

```diff
@@ -1487,77 +1483,71 @@ void lower(const VLS& e, pushpm& i, Vlabel b, size_t z) {
-    v << copy{i.s, d};
-    v << movtqb{d, i.d};
+    v << copy{i.s, i.d};
   });
 }
 
-void lower(const VLS& e, movtdq& i, Vlabel b, size_t z) {
-  lower_impl(e.unit, b, z, [&] (Vout& v) {
-    v << copy{i.s, i.d};
-  });
+void lower(const VLS& e, movzbw& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
 }
 
-void lower(const VLS& e, cmpb& i, Vlabel b, size_t z) {
-  lower_impl(e.unit, b, z, [&] (Vout& v) {
-    auto s0 = v.makeReg();
-    auto s1 = v.makeReg();
-    v << movzbl{i.s0, s0};
-    v << movzbl{i.s1, s1};
-    v << cmpl{s0, s1, i.sf};
-  });
+void lower(const VLS& e, movzbl& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
 }
 
-void lower(const VLS& e, cmpbi& i, Vlabel b, size_t z) {
-  lower_impl(e.unit, b, z, [&] (Vout& v) {
-    auto s1 = v.makeReg();
-    v << movzbl{i.s1, s1};
-    v << cmpli{i.s0, s1, i.sf};
-  });
+void lower(const VLS& e, movzbq& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
 }
 
-#define Y(vasm_opc, conv_opc, load_opc, cmp_opc)            \
-void lower(const VLS& e, vasm_opc& i, Vlabel b, size_t z) { \
-  lower_impl(e.unit, b, z, [&] (Vout& v) {                  \
-    auto s0 = v.makeReg();                                  \
-    v << conv_opc{i.s0, s0};                                \
-    lowerVptr(i.s1, v);                                     \
-    auto s1 = v.makeReg();                                  \
-    v << load_opc{i.s1, s1};                                \
-    v << cmp_opc{s0, s1, i.sf};                             \
-  });                                                       \
+void lower(const VLS& e, movzwl& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
 }
 
-Y(cmpbm, movzbl, loadzbl, cmpl)
-Y(cmpwm, movzwl, loadw, cmpl)
+void lower(const VLS& e, movzwq& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
+}
 
-#undef Y
+void lower(const VLS& e, movzlq& i, Vlabel b, size_t z) {
+  lower_movz(e, i, b, z);
+}
 
-void lower(const VLS& e, testb& i, Vlabel b, size_t z) {
+void lower(const VLS& e, movtdb& i, Vlabel b, size_t z) {
   lower_impl(e.unit, b, z, [&] (Vout& v) {
-    if (i.s0 == i.s1) {
-      v << testbi{(uint8_t)0xff, i.s1, i.sf};
-    } else {
-      auto s0 = v.makeReg();
-      auto s1 = v.makeReg();
-      v << movzbl{i.s0, s0};
-      v << movzbl{i.s1, s1};
-      v << testl{s0, s1, i.sf};
-    }
+    auto d = v.makeReg();
+    v << copy{i.s, d};
+    v << movtqb{d, i.d};
   });
 }
 
-void lower(const VLS& e, testbi& i, Vlabel b, size_t z) {
+void lower(const VLS& e, movtdq& i, Vlabel b, size_t z) {
   lower_impl(e.unit, b, z, [&] (Vout& v) {
-    auto s1 = v.makeReg();
-    v << movzbl{i.s1, s1};
-    v << testli{i.s0, s1, i.sf};
+    v << copy{i.s, i.d};
   });
 }
 
+template<typename load_op, typename cmp_op, typename cmpm>
+void lower_cmpm(const VLS&e, cmpm& i, Vlabel b, size_t z) {
+  lower_impl(e.unit, b, z, [&] (Vout& v) {
+    lowerVptr(i.s1, v);
+    auto tmp0 = v.makeReg();
+    auto tmp1 = v.makeReg();
+    v << copy{i.s0, tmp0};
```

Comment:
The `copy{}`-pattern is used in the functions `lower(movtdb& i)` and `lower(movtdq& i)` above as well for the same reason (in the already existing code). In general the `copy{}` is not harmful, since `optimizeCopies()` cleans up later anyways (you proposed to emit a `copy{}` in `simplify_andi()` so I assume, that relying upon optimizeCopies() is ok).

When doing the proposed Vreg-cast/conversion/proxy trick I get:

```
[ 51%] Building CXX object hphp/runtime/CMakeFiles/hphp_runtime_static.dir/vm/jit/vasm-arm.cpp.o
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp: In instantiation of ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t)::<lambda(HPHP::jit::Vout&)> [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm]’:
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1534:15:   required from ‘struct HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]::<lambda(struct HPHP::jit::Vout&)>’
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1533:13:   required from ‘void HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]’
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1542:39:   required from here
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1537:17: error: call of overloaded ‘Vreg(HPHP::jit::Vreg8&)’ is ambiguous
     v << cmp_op{Vreg(i.s0), tmp, i.sf};
                 ^~~~~~~~~~
In file included from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/arg-group.h:21:0,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-instr.h:21,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-unit.h:26,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-gen.h:24,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:72:
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:59:18: note: candidate: HPHP::jit::Vreg::Vreg(HPHP::jit::Reg8)
   /* implicit */ Vreg(Reg8 r)   : rn(int(r)) {}
                  ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:55:12: note: candidate: HPHP::jit::Vreg::Vreg(size_t)
   explicit Vreg(size_t r) : rn(r) {}
            ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:41:8: note: candidate: constexpr HPHP::jit::Vreg::Vreg(const HPHP::jit::Vreg&)
 struct Vreg {
        ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:41:8: note: candidate: constexpr HPHP::jit::Vreg::Vreg(HPHP::jit::Vreg&&)
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp: In instantiation of ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t)::<lambda(HPHP::jit::Vout&)> [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm]’:
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1534:15:   required from ‘struct HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]::<lambda(struct HPHP::jit::Vout&)>’
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1533:13:   required from ‘void HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]’
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1546:37:   required from here
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1537:17: error: call of overloaded ‘Vreg(HPHP::jit::Vreg16&)’ is ambiguous
     v << cmp_op{Vreg(i.s0), tmp, i.sf};
                 ^~~~~~~~~~
In file included from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/arg-group.h:21:0,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-instr.h:21,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-unit.h:26,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-gen.h:24,
                 from /mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:72:
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:58:18: note: candidate: HPHP::jit::Vreg::Vreg(HPHP::jit::Reg16)
   /* implicit */ Vreg(Reg16 r)  : rn(int(r)) {}
                  ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:55:12: note: candidate: HPHP::jit::Vreg::Vreg(size_t)
   explicit Vreg(size_t r) : rn(r) {}
            ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:41:8: note: candidate: constexpr HPHP::jit::Vreg::Vreg(const HPHP::jit::Vreg&)
 struct Vreg {
        ^~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-reg.h:41:8: note: candidate: constexpr HPHP::jit::Vreg::Vreg(HPHP::jit::Vreg&&)
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1100:6: error: ‘void HPHP::jit::{anonymous}::lower_impl(HPHP::jit::Vunit&, HPHP::jit::Vlabel, size_t, Lower) [with Lower = HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>; size_t = long unsigned int]’, declared using local type ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>’, is used but never defined [-fpermissive]
 void lower_impl(Vunit& unit, Vlabel b, size_t i, Lower lower) {
      ^~~~~~~~~~
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1100:6: error: ‘void HPHP::jit::{anonymous}::lower_impl(HPHP::jit::Vunit&, HPHP::jit::Vlabel, size_t, Lower) [with Lower = HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>; size_t = long unsigned int]’, declared using local type ‘HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm; size_t = long unsigned int]::<lambda(HPHP::jit::Vout&)>’, is used but never defined [-fpermissive]
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1100:6: warning: ‘void HPHP::jit::{anonymous}::lower_impl(HPHP::jit::Vunit&, HPHP::jit::Vlabel, size_t, Lower) [with Lower = HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadw; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpwm]::<lambda(HPHP::jit::Vout&)>]’ used but never defined
/mnt/cmuellner/hhvm/hphp/runtime/vm/jit/vasm-arm.cpp:1100:6: warning: ‘void HPHP::jit::{anonymous}::lower_impl(HPHP::jit::Vunit&, HPHP::jit::Vlabel, size_t, Lower) [with Lower = HPHP::jit::{anonymous}::lower_cmpm(const HPHP::jit::VLS&, cmpm&, HPHP::jit::Vlabel, size_t) [with load_op = HPHP::jit::loadzbl; cmp_op = HPHP::jit::cmpl; cmpm = HPHP::jit::cmpbm]::<lambda(HPHP::jit::Vout&)>]’ used but never defined
hphp/runtime/CMakeFiles/hphp_runtime_static.dir/build.make:11084: recipe for target 'hphp/runtime/CMakeFiles/hphp_runtime_static.dir/vm/jit/vasm-arm.cpp.o' failed
```

When grepping the HHVM source for "Vreg(" to find an example that does something similar, I was not successful to find anything.